http/resource.go

```diff
@@ -157,34 +159,38 @@ var resourcePatchHandler = withUser(func(w http.ResponseWriter, r *http.Request,
 		}
 	}
 
-	if rename {
-		counter := 1
-		dir, name := filepath.Split(dst)
-		ext := filepath.Ext(name)
-		base := strings.TrimSuffix(name, ext)
-
-		for {
-			if _, err = d.user.Fs.Stat(dst); err != nil {
-				break
-			}
-			new := fmt.Sprintf("%s(%d)%s", base, counter, ext)
-			dst = filepath.Join(dir, new)
-			counter++
-		}
-	}
-
 	err = d.RunHook(func() error {
 		switch action {
 		// TODO: use enum
 		case "copy":
 			if !d.user.Perm.Create {
 				return errors.ErrPermissionDenied
 			}
+
+			if rename {
+				dst = AddVersionSuffix(dst, d.user.Fs)
+			}
+
+			err = CheckParent(src, dst)
+			if err != nil {
+				return err
+			}
```

Comment:
I think it could be moved up. Next to the `if dst == "/" || src == "/"` check

Reply:
If there is no problem calling CheckParent() and AddVersionSuffix() before d.user.Perm checking, we can have this:

```go
...
if dst == "/" || src == "/" {
	return http.StatusForbidden, nil
}

override := r.URL.Query().Get("override") == "true"
rename := r.URL.Query().Get("rename") == "true"

if !override && !rename {
	if _, err = d.user.Fs.Stat(dst); err == nil {
		return http.StatusConflict, nil
	}
}

if rename {
	dst = AddVersionSuffix(dst, d.user.Fs)
}

err = CheckParent(src, dst)
if err != nil {
	return err
}

err = d.RunHook(func() error {
...
```
CheckParent() needs to be called after AddVersionSuffix().

Reply:
Why it should be called after AddVersionSuffix? We need to check user input not AddVersionSuffix output

Reply:
Filepath.Rel(src, dst) returns "." when src equals to dst, that causes CheckParent() to return error and makes duplicating files on same folder not possible. I can make it:

1. Call AddVersionSuffix() before CheckParent(), always resulting in src != dst
2. Or have a clause on CheckParent() to not return error when rel == ".".

Reply:
How about
```go
var resourcePatchHandler = withUser(func(w http.ResponseWriter, r *http.Request, d *data) (int, error) {
	src := r.URL.Path
	dst := r.URL.Query().Get("destination")
	action := r.URL.Query().Get("action")
	dst, err := url.QueryUnescape(dst)
	if err != nil {
		return errToStatus(err), err
	}
	src = cleanPath(src)
	dst = cleanPath(dst)
	if err := checkParent(d.user.Fs, src, dst); err != nil {
		return err
	}
	...
}

func checkParent(fs afero.Fs, src, dst string) error {
	isSrcDir, err := afero.IsDir(fs, src)
	if err != nil {
		return err
	}
	if !isSrcDir {
		return nil
	}
	if src != dst && strings.HasPrefix(dst, src) {
		return errors.ErrSourceIsParent
	}
	return nil
}

func cleanPath(fPath string) string {
	return filepath.Clean(filepath.Join("/", fPath))
}
```

Reply:
- HasPrefix() is not good, checkParent() returns error when src="/a/b/c" and dst="/a/b/c2".
- CleanPath() is not needed for the copy action, the fileutils.Copy() will clean the path.
- filepath.Join() causes inconsistency between operational systems.
- For isSrcDir verification to work on checkParent, it is necessary to verify the dst too. rel != "." resolves that.

Other http handlers do not even use filepath.Clean. Since every http handler have its own way of handling path, there will be repetitive code and missing security procedures. The backend should apply a sanitizer for every request, so the http handlers can use a safe and standard path parameter. Maybe in the future i try to review the backend structure to make it more robust.

For now, i think this should be sufficient for this PR:

```go
var resourcePatchHandler = withUser(func(w http.ResponseWriter, r *http.Request, d *data) (int, error) {
	src := r.URL.Path
	dst := r.URL.Query().Get("destination")
	action := r.URL.Query().Get("action")
	dst, err := url.QueryUnescape(dst)
	if err != nil {
		return errToStatus(err), err
	}
	if dst == "/" || src == "/" {
		return http.StatusForbidden, nil
	}
	if err = checkParent(src, dst); err != nil {
		return http.StatusBadRequest, err
	}

	override := r.URL.Query().Get("override") == "true"
	rename := r.URL.Query().Get("rename") == "true"
	if !override && !rename {
		if _, err = d.user.Fs.Stat(dst); err == nil {
			return http.StatusConflict, nil
		}
	}
	if rename {
		dst = addVersionSuffix(dst, d.user.Fs)
	}

	err = d.RunHook(func() error {
		switch action {
		// TODO: use enum
		case "copy":
			if !d.user.Perm.Create {
				return errors.ErrPermissionDenied
			}

			return fileutils.Copy(d.user.Fs, src, dst)
		case "rename":
			if !d.user.Perm.Rename {
				return errors.ErrPermissionDenied
			}
			dst = filepath.Clean("/" + dst)

			return d.user.Fs.Rename(src, dst)
		default:
			return fmt.Errorf("unsupported action %s: %w", action, errors.ErrInvalidRequestParams)
		}
	}, action, src, dst, d.user)

	return errToStatus(err), err
})

func checkParent(src, dst string) error {
	rel, err := filepath.Rel(src, dst)
	if err != nil {
		return err
	}

	rel = filepath.ToSlash(rel)
	if !strings.HasPrefix(rel, "../") && rel != ".." && rel != "." {
		return errors.ErrSourceIsParent
	}

	return nil
}
...
```

Tested on linux and windows.

Reply:
Fair enough